aiida/calculations/plugins/transfer.py

```diff
@@ -0,0 +1,179 @@
+# -*- coding: utf-8 -*-
+###########################################################################
+# Copyright (c), The AiiDA team. All rights reserved.                     #
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+"""Implementation of CalcJobNode to add two numbers for testing and demonstration purposes."""
+import os
+from aiida import orm
+from aiida.engine import CalcJob
+from aiida.common.datastructures import CalcInfo, CodeInfo
+
+
+def validate_transfer_inputs(inputs, ctx):  # pylint: disable=inconsistent-return-statements,unused-argument
+    """Validate the inputs."""
+    source_nodes = inputs['source_nodes']
+    instructions = inputs['instructions']
+
+    if not 'retrieve_files' in instructions.get_dict():
+        errmsg = (
+            'no indication of what to do in the instruction node {}.\n'
+            ' - If you want to retrieve use `retrieve: True`\n'
+            ' - If you want to just gather use `retrieve: False`\n'
+        )
+        return errmsg.format(instructions.uuid)
+
+    if not 'object_list' in instructions.get_dict():
+        errmsg = (
+            'no indication of what to do in the instruction node {}.\n'
+            ' - If you want to retrieve use `retrieve: True`\n'
+            ' - If you want to just gather use `retrieve: False`\n'
+        )
+        return errmsg.format(instructions.uuid)
+    object_list = instructions.get_dict()['object_list']
+
+    source_uuids = [node.pk for node in source_nodes.values()]
+    source_pkids = [node.uuid for node in source_nodes.values()]
+    instruct_ids = [src_node_id for src_node_id, src_locpath, tgt_locpath in object_list]
+
+    for node_id in instruct_ids:
+        its_a_uuid = node_id in source_uuids
+        its_a_pkid = node_id in source_pkids
+        if not (its_a_uuid or its_a_pkid):
+            errmsg = 'node identified by `{}` in instructions has not been passed as an input.'
+            return errmsg.format(node_id)
+
+    for nodename, node in source_nodes.items():  # pylint: disable=unused-variable
+        its_a_uuid = node.uuid in instruct_ids
+        its_a_pkid = node.pk in instruct_ids
+        if not (its_a_uuid or its_a_pkid):
+            errmsg = 'input node provided (pk=`{}`, uuid=`{}`) is not used in instructions.'
+            return errmsg.format(node.pk, node.uuid)
```

Comment:
So, I simplified it a bit, but I think I still preferr when the error says specifically which id was the problem, so I'm still looping twice instead of set comparing. This should be more useful for the user to debug, no?